service/src/main/java/bio/terra/catalog/datarepo/DatarepoService.java

```diff
@@ -47,11 +55,18 @@ public DatarepoService(DatarepoConfiguration datarepoConfig) {
     this.commonHttpClient = new ApiClient().getHttpClient();
   }
 
-  public Map<String, List<String>> getSnapshotIdsAndRoles(AuthenticatedUserRequest user) {
+
+  public Map<String, List<DatasetAccessLevel>> getSnapshotIdsAndRoles(AuthenticatedUserRequest user) {
```

Comment:
If we're going to create an API for roles, it might be more streamlined to define them as a hierarchy (discoverer -> reader -> owner) where `owner` can do everything `reader` can do, and `reader` can do everything `discoverer` can do. Then this could return `Map<String, DatasetAccessLevel>`. If we wanted to support a more complex mix of behaviors, we should use actions / capabilities instead.

Reply:
Hmmm… I agree that makes things easier for permissions checks in the future. I am wondering how to pull that off with the swagger generated enums _and_ backwards compatibility/returning string values.

I think we could do a bigger refactor if we were okay breaking the Catalog UI for a bit while this released, but I think to avoid breaking, we either need to introduce some complexity in the UI, version the endpoints and create a new version (significant back end complexity), accept the subpar mapping, or create a custom class and not use the openAPI generation for this enum.

I lean towards the last one (custom class, no openAPI generation), but am curious your thoughts.